lib/gst/player/gstplayer.c

```diff
@@ -1121,6 +1165,514 @@ element_cb (GstBus * bus, GstMessage * msg, gpointer user_data)
   }
 }
 
+static void
+player_set_flag (GstPlayer *self, gint pos)
+{
+  gint flags;
+
+  g_object_get (self->priv->playbin, "flags", &flags, NULL);
+  flags |= pos;
+  g_object_set (self->priv->playbin, "flags", &flags, NULL);
+}
+
+static void
+player_clear_flag (GstPlayer *self, gint pos)
+{
+  gint flags;
+
+  g_object_get (self->priv->playbin, "flags", &flags, NULL);
+  flags &= ~pos;
+  g_object_set (self->priv->playbin, "flags", &flags, NULL);
+}
+
+typedef struct
+{
+  GstPlayer *player;
+  GstPlayerMediaInfo  *info;
+}MediaInfoUpdatedSignalData;
+
+static gboolean
+media_info_updated_dispatch (gpointer user_data)
+{
+  MediaInfoUpdatedSignalData  *data = user_data;
+
+  g_signal_emit (data->player, signals[SIGNAL_MEDIA_INFO_UPDATED], 0, data->info);
+
+  return FALSE;
+}
+
+static void
+free_media_info_updated_signal_data (MediaInfoUpdatedSignalData *data)
+{
+  GST_DEBUG_OBJECT (data->player, "media_info %p", data->info);
+  g_object_unref (data->info);
+  g_slice_free (MediaInfoUpdatedSignalData, data);
+}
+
+/*
+ * emit_media_info_updated_signal:
+ *
+ * create a new copy of self->priv->media_info object and emits the newly created
+ * copy to user application. The newly created media_info will be unref'ed
+ * as part of signal finalize method.
+ */
+static void
+emit_media_info_updated_signal (GstPlayer *self)
+{
+  if (self->priv->dispatch_to_main_context) {
+    MediaInfoUpdatedSignalData *data = g_slice_new
+                                        (MediaInfoUpdatedSignalData);
+    data->player = self;
+    g_mutex_lock (&self->priv->lock);
+    data->info = gst_player_media_info_copy (self->priv->media_info);
+    g_mutex_unlock (&self->priv->lock);
+
+    GST_DEBUG_OBJECT (self, "media_info %p", data->info);
+
+    g_main_context_invoke_full (self->priv->application_context,
+        G_PRIORITY_DEFAULT, media_info_updated_dispatch,
+        data, (GDestroyNotify) free_media_info_updated_signal_data);
+  } else {
+    g_signal_emit (self, signals[SIGNAL_MEDIA_INFO_UPDATED], 0);
```

Comment:
You have to provide the info in g_signal_emit() here too. This should be: lock(); info=copy_info(); unlock(); emit(); unref(info);